Simplify batching logic in PendingStateManager

kian-thompson opened the pull request:
# [AB#1833](https://dev.azure.com/fluidframework/internal/_workitems/edit/1833)

## Description

Batch tracking inside the `PendingStateManager` will now be done through the batch metadata on ops to indicate starts and ends of batches.

Before, we were using `flush` calls to indicate a batch had ended because the `batch: false` metadata would not be updated in messages in the `PendingStateManager`. Due to changes in the batching workflow in `ContainerRuntime`, the `batch: false` metadata is available.

## Does this introduce a breaking change?

`IPendingFlush` has been deprecated. Use batch metadata on `IPendingMessage` instead to indicate the end of a batch.

kian-thompson committed:
Initial work to remove flush tracking

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
Simplify some logic
- batch begin and end
- replaying pending states

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
Fix test

kian-thompson committed:
Write old local state format (in case of rollback)

kian-thompson committed:
Temporary fix for testing
- main fix is https://github.com/microsoft/FluidFramework/pull/12581

kian-thompson committed:
Adjustments to APIs

kian-thompson committed:
Add unit test coverage for conversion of local state

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
merge next

kian-thompson commented:
@dannimad FYI, this might affect the work you're doing around pending local state

kian-thompson committed:
Only flush orderSequentially when not TurnBased

andre4i commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -347,80 +335,52 @@ export class PendingStateManager implements IDisposable {
+                    pendingMessage.localOpMetadata,
+                    pendingMessage.opMetadata);
             }
             pendingStatesCount--;
```
I might be misunderstanding this, but is should `pendingStatesCount--` be inside the else statement? Isn't this double decrementing in case of batch messages?

kian-thompson replied:
The logic is technically correct as-is, but it is definitely confusing. I rearranged some of it to hopefully be clearer about checks.

kian-thompson committed:
Make onFlush functional while it's deprecated

kian-thompson committed:
Andrei codereview

kian-thompson committed:
Merge branch 'next' into 1833-pending-state-manager-batching

kian-thompson committed:
Label future work with ADO item

agarwal-navin commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -347,80 +341,55 @@ export class PendingStateManager implements IDisposable {
+                assert(pendingMessagesCount > 0, "Last pending message cannot be a batch begin");
+
+                this.stateHandler.orderSequentially(() => {
+                    while (pendingMessagesCount >= 0) { // check is >= because batch end may be last pending message
```
There should probably be validation inside this loop that we do infact get a batch end. Otherwise, if this happens it would be very hard to debug.

agarwal-navin commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -151,23 +188,17 @@ export class PendingStateManager implements IDisposable {
 
     /**
-     * Called when flush() is called on the ContainerRuntime to manually flush messages.
+     * @deprecated Use batch metadata to indicate end of batch. To be removed in 2.0.0-internal.4.0.0 (AB#2496)
```
Why do we need back-compat here? Aren't all the call sites in container runtime which is in the same layer? Is it because the change to add batch metadata to messages is behind a feature flag?

kian-thompson replied:
Andre mentioned this function was only deprecated and not removed (before, I made this method do nothing at all). So, I made this intermediary step if somebody still wants to use the `onFlush` method.

Looking at it more, this class isn't exported. So, it's probably fine if this method loses its functionality as no one can extend it?

kian-thompson replied:
@agarwal-navin FYI, I don't think adding the batch metadata is part of a feature flag? There isn't any check in `ContainerRuntime.flushBatch(...)`

kian-thompson committed:
Add extra assert for batch validation in replayPendingStates

kian-thompson committed:
Adjust breaking/deprecated items
- publicly export interface remains deprecated
- internal API changes removed as they are non-breaking
